### PR TITLE
(259) Blog Topic tests

### LINKED ISSUE #259: CMS: Adding Blog Topic to verify "Categories" in Blog Series and Blog Posts
## As a cms user I want to be able to create Blog Topic so that I can see Categories on the Blog Series and Blog Posts.

Create a new blog topic (under Structure > Taxonomy > Blog Topics > List Terms > Add Terms) 

Fill in the description, to display at the top of the blog topic page (using blog post intro text styling) 

Ensure taxonomy term only displays as Category link in right rail when posts are tagged to the topic 

> **NOTE** : We will be adding the Blog Topic test scenarios on the $BlogSeries.feature and implement it on $BlogSeries -> index.js.     Updated $BlogSeries.feature with Blog Topic test scenarios.

> **ESTIMATE** TBD

### Acceptance Criteria

```gherkin
Feature: As a cms user I want to be able to create Blog Series content type to promote Blog Series and Blog Topic.

    Scenario: User is adding new Blog Series content type
        Given user is navigating to "/user/login"
        When user enters credentials
        And user clicks "Log in" button
        Then user is logged in and the user name "admin" is displayed in the toolbar
        And the tool bar appears at the top
        When user clicks on "Content" tab
        And user clicks on "Add content" action button
        And user clicks on "Blog Series" content type
        Then page title is "Create Blog Series"
        When user selects test site section
        And user fills out the following fields
            | fieldLabel                 | value                                             | field_name                       |
            | Page Title                 | Automated Test Blog Series                        | title                            |
            | Pretty URL                 | blog-series                                       | field_pretty_url                 |
            | Browser Title              | Automated Test Blog Series - Browser Title        | field_browser_title              |
            | Card Title                 | Automated Test Blog Series - Card Title           | field_card_title                 |
            | Meta Description           | Automated Blog Series Meta Description            | field_page_description           |
            | Feature Card Description   | Automated Test Blog Series - Feature Card Desc    | field_feature_card_description   |
            | Blog Series Shortname      | TestBlogSeries                                    | field_blog_series_shortname      |
            | Subscription URL           | https://feedburner-test.cancer.gov/testblogseries | field_feedburner_url             |
            | Recommended Content Header | Recommended Text                                  | field_recommended_content_header |
        And browser waits
        And user uploads test file "heroimagenewsdesktop.jpg"
        And system waits for file upload process
        And user enters "Text Alt" into the "Alternative text" text field
        And user enters "Blog Series List Description" into "List Description" text field
        And user fills out "About Blog" text area with "This is all about Blog Series." in the blog series
        And user selects "Published" from "Save as" dropdown
        Then user saves the content page

    Scenario: Verify newly created content Blog Series
        Given user is navigating to the front end site with path site section plus "blog-series"
        And the banner image is displayed
        Then page title is "Automated Test Blog Series"
        And blog series description reads "This is all about Blog Series"
        And the "Subscribe" link appears with url "https://feedburner-test.cancer.gov/testblogseries"
        And the "Archive" accordion is displayed
        And blog posts list does not appear in the archive
        And the page contains meta tags with the following names
            | name        | content                                |
            | description | Automated Blog Series Meta Description |

    Scenario: User is adding new Blog Post content type without any image so it can be added to Blog Series
        Given user is navigating to "/user/login"
        When user enters credentials
        And user clicks "Log in" button
        Then user is logged in and the user name "admin" is displayed in the toolbar
        And the tool bar appears at the top
        When user clicks on "Content" tab
        And user clicks on "Add content" action button
        And user clicks on "Blog Post" content type
        Then page title is "Create Blog Post"
        And user selects "Automated Test Blog Series" from Blog Series dropdown
        And browser waits
        And user fills out the following fields
            | fieldLabel | value                      | field_name       |
            | Pretty URL | blog-post-test-blog-series | field_pretty_url |
        And browser waits
        And user fills out the following fields
            | fieldLabel               | value                                            | field_name                     |
            | Page Title               | Automated Test Blog Post for testing Blog Series | title                          |
            | Browser Title            | Automated Test Blog Post - Browser Title         | field_browser_title            |
            | Card Title               | Automated Test Blog Post - Card Title            | field_card_title               |
            | Meta Description         | Automated Blog Post Meta Description             | field_page_description         |
            | Feature Card Description | Automated Test Blog Post - Feature Card Desc     | field_feature_card_description |
            | Author                   | Automated Test Blog Post Author                  | field_author                   |
        And user enters "Blog Post List Description" into "List Description" text field
        And user enters "Blog Post Intro Text" as intro text
        And user fills out "Body" text area with "This is a description of Blog Post content type."
        And user selects "Published" from "Save as" dropdown
        Then user saves the content page

    Scenario: User is adding Blog Topic to the already created Blog Series
        Given user is navigating to "/user/login"
        When user enters credentials
        And user clicks "Log in" button
        Then user is logged in and the user name "admin" is displayed in the toolbar
        And the tool bar appears at the top
        When user clicks on "Structure" tab
        And user clicks on "Taxonomy" sub tab
        And user selects "List terms" option from Operations for "Blog Topics"
        And user clicks on "Add term" action button
        Then page title is "Add term"
        And user fills out the following fields
            | fieldLabel | value           | field_name                       |
            | Name       | Test Blog Topic | name[0][value]                   |
            | Pretty URL | test-blog-topic | field_topic_pretty_url[0][value] |
        And user selects "Automated Test Blog Series" radio button under "Owner Blog"
        And browser waits
        And user enters "Blog Topic Description" into "Description" field
        Then user saves the content page

    Scenario: Edit and republish Blog Post content type
        Given user is navigating to "/user/login"
        When user enters credentials
        And user clicks "Log in" button
        Then user is logged in and the user name "admin" is displayed in the toolbar
        And the tool bar appears at the top
        When user clicks on "Content" tab
        And user clicks on title with url "blog-post-test-blog-series" under "/about-cancer/understanding/blog-series" from the list of content
        And user clicks on the tool bar status green button "Published"
        And user clicks "View in edit form" button from other actions
        And user checks "Test Blog Topic" checkbox
        And user selects "Do Not Display" from "Public Use Text" dropdown
        When user saves the content page
        And user clicks on the tool bar status red button "Editing"
        And user selects "Quick Publish" from workflow actions

    Scenario: Edit and republish Blog Series content type
        Given user is navigating to "/user/login"
        When user enters credentials
        And user clicks "Log in" button
        Then user is logged in and the user name "admin" is displayed in the toolbar
        And the tool bar appears at the top
        When user clicks on "Content" tab
        And user clicks on title with url "blog-series" from the list of content
        And user clicks on the tool bar status green button "Published"
        And user clicks "View in edit form" button from other actions
        And user clears out "Pretty URL" field
        And user clears out "Page Title" field
        And user clears out "Meta Description" field
        And user fills out the following fields
            | fieldLabel       | value                                              | field_name             |
            | Pretty URL       | blog-series-edited                                 | field_pretty_url       |
            | Page Title       | Automated Test Blog Series Edited                  | title                  |
            | Meta Description | Automated Test Blog Series Meta Description Edited | field_page_description |
        And user clicks on dropdown button toggle to view Featured Posts
        And user clicks on "Select entities" button to link to the Featured Post
        Then "Select entities" section appears
        And user selects "Automated Test Blog Post for testing Blog Series" item from the list
        And user clicks on "Add Blog content" button to select item
        And browser waits
        And user clicks on "Select entities" button to link to the Featured Post
        Then "Select entities" section appears
        And user selects "Automated Test Blog Post for testing Blog Series" item from the list
        And user clicks on "Add Blog content" button to select item
        And browser waits
        And user clicks on "Select entities" button to link to the Featured Post
        Then "Select entities" section appears
        And user selects "Automated Test Blog Post for testing Blog Series" item from the list
        And user clicks on "Add Blog content" button to select item
        And browser waits
        And user selects 1 Promotional Image from the list of images for blog series
        And user remembers the source of selected promo image for blog series for further verification
        And user selects "Do Not Display" from "Public Use Text" dropdown
        When user saves the content page
        And user clicks on the tool bar status green button "Editing"
        And user selects "Quick Publish" from workflow actions

    Scenario: User is adding new Blog Post content type with images so it can be added to Blog Series
        Given user is navigating to "/user/login"
        When user enters credentials
        And user clicks "Log in" button
        Then user is logged in and the user name "admin" is displayed in the toolbar
        And the tool bar appears at the top
        When user clicks on "Content" tab
        And user clicks on "Add content" action button
        And user clicks on "Blog Post" content type
        Then page title is "Create Blog Post"
        And user selects "Automated Test Blog Series Edited" from "Blog Series" dropdown
        And browser waits
        And user fills out the following fields
            | fieldLabel               | value                                             | field_name                     |
            | Pretty URL               | blog-post2-test-blog-series                       | field_pretty_url               |
            | Page Title               | Automated Test Blog Post2 for testing Blog Series | title                          |
            | Browser Title            | Automated Test Blog Post2 - Browser Title         | field_browser_title            |
            | Card Title               | Automated Test Blog Post2 - Card Title            | field_card_title               |
            | Meta Description         | Automated Blog Post2 Meta Description             | field_page_description         |
            | Feature Card Description | Automated Test Blog Post2 - Feature Card Desc     | field_feature_card_description |
            | Author                   | Automated Test Blog Post2 Author                  | field_author                   |
        And user enters "Blog Post List Description" into "List Description" text field
        And user enters "Blog Post Intro Text" as intro text
        And user fills out "Body" text area with "This is a description of Blog Post content type."
        And user selects 1 Lead Image from the list of images for blog series
        And browser waits
        And user remembers the source of selected lead image for blog series for further verification
        And user selects 2 Promotional Image from the list of images for blog series
        And browser waits
        And user remembers the source of selected promo image for blog series for further verification
        And user selects "Published" from "Save as" dropdown
        Then user saves the content page

    Scenario: Verify edited content and Blog Topic in the Blog Series
        Given user is navigating to the front end site with path site section plus "blog-series-edited"
        Then page title is "Automated Test Blog Series Edited"
        And the "Featured Posts" managed list appears
        And public use text is not displayed
        And the Featured Posts section is displayed with the following cards
            | title                                 | linkEnding | expectedDate | author                          |
            | Automated Test Blog Post - Card Title | /blog-post | today's date | Automated Test Blog Post Author |
            | Automated Test Blog Post - Card Title | /blog-post | today's date | Automated Test Blog Post Author |
            | Automated Test Blog Post - Card Title | /blog-post | today's date | Automated Test Blog Post Author |
        And the page contains meta tags with the following names
            | name        | content                                            |
            | description | Automated Test Blog Series Meta Description Edited |
        Then list of blog post has the following posts
            | title                                             | url                                                                                         | expectedDate | author                           |
            | Automated Test Blog Post for testing Blog Series  | {TEST_SITE_SECTION}/blog-series-{RANDOM}/{YEAR}/blog-post-test-blog-series-{RANDOM}         | today's date | Automated Test Blog Post Author  |
            | Automated Test Blog Post2 for testing Blog Series | {TEST_SITE_SECTION}/blog-series-edited-{RANDOM}/{YEAR}/blog-post2-test-blog-series-{RANDOM} | today's date | Automated Test Blog Post2 Author |
        And the promotional image is matching the earlier selected image
        And the Continue Reading link appears with the following href
            | linkName           | linkHref                                                                                    |
            | Continue Reading > | {TEST_SITE_SECTION}/blog-series-{RANDOM}/{YEAR}/blog-post-test-blog-series-{RANDOM}         |
            | Continue Reading > | {TEST_SITE_SECTION}/blog-series-edited-{RANDOM}/{YEAR}/blog-post2-test-blog-series-{RANDOM} |
        And the "Categories" managed list appears without the date
        And the Test Blog Topic link appears with the following href
            | linkName        | linkHref                                                                       |
            | Test Blog Topic | {TEST_SITE_SECTION}/blog-series-edited-{RANDOM}?topic=test-blog-topic-{RANDOM} |

    Scenario: Verify Blog Topic at the front end on Blog Post without images content type
        Given user is navigating to the blog "blog-post-test-blog-series" under "about-cancer/understanding/blog-series"
        And the "Categories" managed list appears without the date
        And the Test Blog Topic link appears with the following href
            | linkName        | linkHref                                                                       |
            | Test Blog Topic | {TEST_SITE_SECTION}/blog-series-edited-{RANDOM}?topic=test-blog-topic-{RANDOM} |

    Scenario: Verify Blog Topic at the front end on Blog Post with images content type
        Given user is navigating to the blog "blog-post2-test-blog-series" under "/about-cancer/understanding/blog-series-edited"
        And the "Categories" managed list appears without the date
        And the Test Blog Topic link appears with the following href
            | linkName        | linkHref                                                                       |
            | Test Blog Topic | {TEST_SITE_SECTION}/blog-series-edited-{RANDOM}?topic=test-blog-topic-{RANDOM} |

    Scenario: Clean up
        Given user is navigating to "/user/login"
        When user enters credentials
        And user clicks "Log in" button
        Then user is logged in and the user name "admin" is displayed in the toolbar
        And the tool bar appears at the top
        When user clicks on "Content" tab
        And user selects a checkbox next to title "Automated Test Blog Series Edited" with url "blog-series-edited" from the list of content
        And user selects a checkbox next to title "Automated Test Blog Post for testing Blog Series" with url "blog-post-test-blog-series" from the list of content
        And user selects a checkbox next to title "Automated Test Blog Post2 for testing Blog Series" with url "blog-post2-test-blog-series" from the list of content
        And user clicks on the "Apply to selected items" action button
        Then page title is "Are you sure you want to delete these content items?"
        When user clicks on "Delete" button
        Then the confirmation text "Deleted 3 content items." appears on a screen
        And the content item with url "blog-series-edited" does not exist in the list of content
        And the content item with url "blog-post-test-blog-series" does not exist in the list of content
        And the content item with url "blog-post2-test-blog-series" does not exist in the list of content
        When user clicks on "Structure" tab
        And user clicks on "Taxonomy" sub tab
        And user selects "List terms" option from Operations for "Site Sections"
        And user clicks on "Delete" from  dropdown button under "Blog Topic"
        Then page title is "Are you sure you want to delete the taxonomy term Test Blog Topic?"
        When user clicks on "Delete" button
        Then the confirmation text "Deleted term Test Blog Topic." appears on a screen


```

### Resources:
* Feature: JIRA Ticket #
* Mockups: *Here goes a URL to or the name of the mockup(s) in inVision*
* Web Page: *Here goes a URL to an existing web page*

### Notes

*Some complementary notes if necessary:*

* > Here goes a quote from an email
* Here goes whatever useful information can exist…

---

### Prerequisites
- *required dependency for identifying the problem, scope and then solution for this issue*
- [ ] Short Spike Description - Issue #9999

### Tasks
- [ ] Short Task Description - Issue #9999
- [ ] Short Task Description - Issue #9999

### Technical Notes
- More notes, or technical nuances/direction

### PR DESCRIPTION
**Pull request by alinai**
Closes #259 

We will be adding the Blog Topic test scenarios on the $BlogSeries.feature and implement it on $BlogSeries.feature -> index.js.
Updated $BlogSeries.feature with Blog Topic test scenarios.

### CODE REVIEW

**Comment by alinai** [comment]
cypress test run on dev-acsf:
image

**Comment by alinai** [comment]
npm test run on dev-acsf:

image

**Comment by alinai** [comment]
cypress test run on dev-ac:

image

File: cypress/e2e/CMS/$BlogSeries/index.js [hunk 1]
```diff
@@ -111,6 +124,30 @@ And('blog series description reads {string}', (desc) => {
     cy.get(`p:contains('${desc}')`).should('be.visible');
 });
 
+And('user clicks on {string} sub tab', (contentSubTab) => {
+    cy.get(`ul.admin-list li a:contains("${contentSubTab}")`).click({ force: true });
+});
+
+And('user selects {string} option from Operations for {string}', (option, label) => {
+    cy.get('[data-drupal-selector="edit-vocabularies-cgov-blog-topics"] > :nth-child(4) > .dropbutton-wrapper > .dropbutton-widget > .dropbutton > .list > a').click({ force: true });
```
Comment (KateMashkinaNIH) [inline cypress/e2e/CMS/$BlogSeries/index.js:132]: this locator looks very brittle - could you try something like
cy.get(a:contains('${label}').parent().parent().find (a[href*="cgov_blog_topics"]).contains(option).click()

File: cypress/e2e/CMS/$BlogSeries/index.js [hunk 2]
```diff
@@ -267,4 +319,12 @@ And('user selects a checkbox next to title {string} with url {string} from the l
 
 And('user clicks on the {string} action button', (buttonLabel) => {
     cy.get(`input[value='${buttonLabel}']`).first().click();
+});
+
+And('user clicks on the {string} action button', (buttonLabel) => {
+    cy.get(`input[value='${buttonLabel}']`).first().click();
```
Comment (KateMashkinaNIH) [inline cypress/e2e/CMS/$BlogSeries/index.js:325]: duplicate step from above

File: cypress/e2e/CMS/$BlogSeries/index.js [hunk 3]
```diff
@@ -111,6 +124,30 @@ And('blog series description reads {string}', (desc) => {
...
+    cy.get('fieldset#edit-field-blog-topics--wrapper input').parent().find(`label:contains("${checkboxLbl}")`).as('checkbox');
+    cy.get('@checkbox').parent().find('input').check();
+});
+
+And('user clicks on the tool bar status red button {string}', (status) => {
+    cy.get(`a[data-label='${status}']`).click();
+});
+
+And('user enters {string} into {string} field', (desc, label) => {
+    cy.getIframeBody("iframe[title='Rich Text Editor, Description field']").find('p').type(desc);
```
Comment (KateMashkinaNIH) [inline cypress/e2e/CMS/$BlogSeries/index.js:149]: label variable is not used

File: cypress/e2e/CMS/$BlogSeries/index.js [hunk 4]
```diff
@@ -111,6 +124,30 @@ And('blog series description reads {string}', (desc) => {
...
+    cy.get(`#edit-field-owner-blog label:contains("${option}")`).first().click();
+});
+
+And('user checks {string} checkbox', (checkboxLbl) => {
+    cy.get('fieldset#edit-field-blog-topics--wrapper input').parent().find(`label:contains("${checkboxLbl}")`).as('checkbox');
+    cy.get('@checkbox').parent().find('input').check();
+});
+
+And('user clicks on the tool bar status red button {string}', (status) => {
+    cy.get(`a[data-label='${status}']`).click();
```
Comment (KateMashkinaNIH) [inline cypress/e2e/CMS/$BlogSeries/index.js:145]: so this has the same code implementation as 2 more steps - I suggest re-using step in common 
user clicks on the tool bar status green button {string} - the color doesnt matter, the status is! and it is a variable, so we are good 

also, remove existing steps that say red and yellow button respectfully and update them to be green instead

File: cypress/e2e/CMS/$BlogSeries/index.js [hunk 5]
```diff
@@ -111,6 +124,30 @@ And('blog series description reads {string}', (desc) => {
     cy.get(`p:contains('${desc}')`).should('be.visible');
 });
 
+And('user clicks on {string} sub tab', (contentSubTab) => {
+    cy.get(`ul.admin-list li a:contains("${contentSubTab}")`).click({ force: true });
```
Comment (KateMashkinaNIH) [inline cypress/e2e/CMS/$BlogSeries/index.js:128]: this step exist in common - remove it here

**Review by KateMashkinaNIH — CHANGES_REQUESTED** [review]

File: cypress/e2e/CMS/$BlogSeries/index.js (hunk 1 above)
Comment (alinai) [reply cypress/e2e/CMS/$BlogSeries/index.js:132]: Updated

**Comment by alinai** [comment]
npm test run on dev-acsf:
image

cypress test run on dev-ac:
image

cypress test run on dev-acsf:
image

File: cypress/e2e/CMS/$BlogSeries/index.js (hunk 2 above)
Comment (alinai) [reply cypress/e2e/CMS/$BlogSeries/index.js:325]: Removed the duplicated step

File: cypress/e2e/CMS/$BlogSeries/index.js (hunk 3 above)
Comment (alinai) [reply cypress/e2e/CMS/$BlogSeries/index.js:149]: Fixed the step in the feature file.

File: cypress/e2e/CMS/$BlogSeries/index.js (hunk 4 above)
Comment (alinai) [reply cypress/e2e/CMS/$BlogSeries/index.js:145]: Fixed by updating the feature file and removing the code from index.js

File: cypress/e2e/CMS/$BlogSeries/index.js (hunk 5 above)
Comment (alinai) [reply cypress/e2e/CMS/$BlogSeries/index.js:128]: The step fails with the locator in common . So we are keeping the existing code in index.js for this step.

**Comment by alinai** [comment]
@KateMashkinaNIH , Updated per review. Thanks!

**Review by KateMashkinaNIH — APPROVED** [review]

**Pull request merged**